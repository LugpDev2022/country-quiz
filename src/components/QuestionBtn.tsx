interface Props {
  questionNumber: number;
  active?: boolean;
}

const QuestionBtn: React.FC<Props> = ({ questionNumber, active = false }) => {
  return (
    <a href='#' className={active ? 'question-btn-active' : 'question-btn'}>
      {questionNumber}
    </a>
  );
};

export default QuestionBtn;
