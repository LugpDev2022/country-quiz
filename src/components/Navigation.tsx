import { useLocation } from 'react-router-dom';
import queryString from 'query-string';

import QuestionBtn from './QuestionBtn';

const numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];

const Navigation = () => {
  const { search } = useLocation();

  const { question } = queryString.parse(search) as { question: string };

  return (
    <nav className='flex flex-wrap gap-3 justify-center mb-10'>
      {numbers.map((number) => (
        <QuestionBtn
          key={number}
          questionNumber={number}
          active={parseInt(question) === number}
        />
      ))}
    </nav>
  );
};

export default Navigation;
