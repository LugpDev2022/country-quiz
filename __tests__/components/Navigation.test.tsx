import React, { useContext } from 'react';
import { render } from '@testing-library/react';
import { vi } from 'vitest';

import Navigation from '../../src/components/Navigation';
import { MemoryRouter } from 'react-router';
import QuestionBtn from '../../src/components/QuestionBtn';

// Mock useContext from React to control the context values used in the component
vi.mock('react', async (importOriginal) => {
  const actual = await importOriginal<typeof import('react')>();
  return {
    ...actual,
    useContext: vi.fn(),
  };
});

// Mock QuestionBtn component
vi.mock('../../src/components/QuestionBtn', () => {
  return {
    default: vi.fn(() => null), // Ensure it returns null to avoid rendering issues
  };
});

describe('<Navigation /> Tests', () => {
  beforeEach(() => {
    vi.clearAllMocks();
    // Reset useContext mock to default state before each test
    useContext.mockReturnValue({
      currentQuestionNumber: 2,
      questionsData: [
        { question: 'Test question', selectedAnswer: null },
        { question: 'Test question 2', selectedAnswer: null },
      ],
    });
  });

  test('should match snapshot', () => {
    const { container } = render(
      <MemoryRouter>
        <Navigation />
      </MemoryRouter>
    );
    expect(container).toMatchSnapshot();
  });

  test('should call <QuestionBtn /> with the correct props', () => {
    render(
      <MemoryRouter>
        <Navigation />
      </MemoryRouter>
    );

    expect(QuestionBtn).toHaveBeenNthCalledWith(
      1,
      { active: false, questionNumber: 1 },
      {}
    );

    expect(QuestionBtn).toHaveBeenNthCalledWith(
      2,
      { active: true, questionNumber: 2 },
      {}
    );
  });

  test('should mark question button as active when answered', () => {
    useContext.mockReturnValue({
      currentQuestionNumber: 1,
      questionsData: [
        { question: 'Test question', selectedAnswer: null },
        { question: 'Test question 2', selectedAnswer: 2 },
      ],
    });

    render(
      <MemoryRouter>
        <Navigation />
      </MemoryRouter>
    );

    expect(QuestionBtn).toHaveBeenNthCalledWith(
      2,
      { active: true, questionNumber: 2 },
      {}
    );
  });
});
