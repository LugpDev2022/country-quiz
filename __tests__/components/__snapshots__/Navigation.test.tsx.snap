// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Navigation /> Tests > should match snapshot 1`] = `
<div>
  <nav
    class="flex flex-wrap gap-3 justify-center mb-10"
  />
</div>
`;
